plugins/lookup/etcd3.py

```diff
@@ -0,0 +1,230 @@
+            - name: ETCDCTL_USER
+            type: str
+        password:
+            description:
+            - auth user password
+            default: None
+            env:
+            - name: ETCDCTL_PASSWORD
+            type: str
+
+    requirements:
+    - "etcd3 >= 0.10"
+'''
+
+EXAMPLES = '''
+    - name: "a value from a locally running etcd"
+      debug:
+        msg: "{{ lookup('community.general.etcd3', 'foo/bar') }}"
+
+    - name: "values from multiple folders on a locally running etcd"
+      debug:
+        msg: "{{ lookup('community.general.etcd3', 'foo', 'bar', 'baz') }}"
+
+    - name: "look for a key prefix"
+      debug:
+        msg: "{{ lookup('community.general.etcd3', '/foo/bar', prefix=True) }}"
+
+    - name: "connect to etcd3 with a client certificate"
+      debug:
+        msg: "{{ lookup('community.general.etcd3', 'foo/bar', cert_cert='/etc/ssl/etcd/client.pem', cert_key='/etc/ssl/etcd/client.key') }}"
+'''
+
+RETURN = '''
+    _raw:
+        description:
+        - List of keys and associated values.
+        type: list
+        elements: complex
+'''
+
+import os
+import re
+import traceback
+
+from ansible.errors import AnsibleLookupError
+from ansible.utils.display import Display
+from ansible.plugins.lookup import LookupBase
+from ansible.module_utils.basic import missing_required_lib
+from ansible.module_utils._text import to_native
+
+try:
+    import etcd3
+    HAS_ETCD = True
+except ImportError:
+    HAS_ETCD = False
+
+display = Display()
+
+
+def etcd3_client(client_params):
+    etcd = None
+    try:
+        etcd = etcd3.client(**client_params)
+        etcd.status()
+    except Exception as exp:
+        raise AnsibleLookupError('Cannot connect to etcd cluster: %s' % (to_native(exp)))
+    return etcd
+
+
+class LookupModule(LookupBase):
+
+    def log_env_override(self, params, key):
+        display.verbose(
+            "Overriding etcd3 '{0}' param with ENV variable '{1}'".format(
+                to_native(key),
+                to_native(params[key])
+            )
+        )
+
+    def run(self, terms, variables, **kwargs):
+
+        self.set_options(var_options=variables, direct=kwargs)
+
+        if not HAS_ETCD:
+            display.error(missing_required_lib('etcd3'))
+            return None
+
+        # get etcd3 optional args
+        client_params = {
+            'host': '127.0.0.1',
+            'port': 2379,
+            'ca_cert': None,
+            'cert_cert': None,
+            'cert_key': None,
+            'timeout': 60,
+            'user': None,
+            'password': None
+        }
+
+        # use ETCD environment variables if they are set
+        etcd3_envs = {
+            'ca_cert': 'ETCDCTL_CACERT',
+            'cert_cert': 'ETCDCTL_CERT',
+            'cert_key': 'ETCDCTL_KEY',
+            'user': 'ETCDCTL_USER',
+            'password': 'ETCDCTL_PASSWORD',
+            'timeout': 'ETCDCTL_DIAL_TIMEOUT',
+        }
+
+        for key in etcd3_envs.keys():
```

Comment:
Why is this loop needed? Ansible does take care of the environment variables, since you put them in the documentation above.

Reply:
Also, this does prevent the user overwriting the env variables by specifying a option to the lookup plugin.